interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -287,6 +299,29 @@ func (pdb *PromptDB) AddOrMerge(metadata *prompting.Metadata, path string, reque
 			ids: make(map[prompting.IDType]int),
 		}
 		userEntry = pdb.perUser[metadata.User]
+		userEntry.expirationTimer = time.AfterFunc(initialTimeout, func() {
+			pdb.mutex.Lock()
+			expiredPrompts := userEntry.prompts
+			// Clear all outstanding prompts for the user
+			userEntry.prompts = nil
+			userEntry.ids = make(map[prompting.IDType]int)
+			// Restart expiration timer while holding the lock, so we don't
+			// overwrite a newly-set activity timeout with an initial timeout.
+			// With the lock held, no activity can occur, so no activity timeout
+			// can be set.
+			userEntry.expirationTimer.Reset(initialTimeout)
```

Comment:
This and the other code which resets the timer is gated by the mutex. so I'm thinking there's a scenario possible that the timer expires and the after-func is scheduled to run in a goroutine, but then is held by the lock as there's an ongoing call to query/reply which also holds the lock, now the call ends, the lock is released and after-func proceeds to drop all prompts despite activity from the user. Perhaps we should check here whether Reset() returns true, which would likely indicate that the timer is active again, and simply exit without dropping the prompts and there's an evidence of user activity?

Reply:
I think you're right. The original reasoning is that if the timer expires, we should expire all prompts, regardless of if there's a race with a reply or prompt detail retrieval. But if the prompt or retrieval goes through, the client should expect that the prompts don't immediately expire. So you're right, the timer should be reset to `activityTimeout` and the function should return without expiring prompts.